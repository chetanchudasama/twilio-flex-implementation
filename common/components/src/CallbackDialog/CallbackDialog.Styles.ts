import styled from "react-emotion";

export const CallbackDialogStyles = styled("div")({
  padding: "10px",
  ".reason-field": {
    paddingTop: "15px",
  },
});
