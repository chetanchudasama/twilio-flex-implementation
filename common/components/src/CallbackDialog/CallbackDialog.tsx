import { Grid, TextField } from "@material-ui/core";
import React, { useState } from "react";
import { ResponsiveDialog } from "../ResponsiveDialog/index";
import { CallbackDialogStyles } from "./CallbackDialog.Styles";
import { MuiDatePicker } from "../MuiDatePicker/MuiDatePicker";

export interface CallbackDialogProps {
  open: boolean;
  noteOptional?: boolean;
  callbackBooked?: Date;
  setCallbackDetail: (callback: Date | null, note?: string) => void;
  handleCloseDialog: () => void;
}

export const CallbackDialog: React.FC<CallbackDialogProps> = ({
  open,
  noteOptional,
  callbackBooked,
  setCallbackDetail,
  handleCloseDialog,
}) => {
  const [openConfirmationDialog, setOpenConfirmationDialog] = useState(false);
  const [callbackDateTime, setCallbackDateTime] = useState<Date | null>(
    callbackBooked as Date
  );
  const [reason, setReason] = useState<string>("");
  const [highlightRequiredDetail, setHighlightRequiredDetail] =
    useState<boolean>(false);

  const handleChange = (value: Date | string, propertyName: string) => {
    if (propertyName === "callbackDateTime") {
      setCallbackDateTime(value as Date);
    } else {
      setReason(value as string);
    }
  };

  const handleSetCallback = () => {
    setHighlightRequiredDetail(true);
    if (callbackDateTime && (!noteOptional ? reason : true)) {
      setCallbackDetail(callbackDateTime, reason);
    }
  };

  return (
    <>
      {open && (
        <ResponsiveDialog
          title={callbackBooked ? "Update Callback" : "Set Callback"}
          open={open}
          onCancel={() => handleCloseDialog()}
          onConfirm={() => handleSetCallback()}
          maxWidth="xs"
          showActionButtons
          cancelText="close"
          okText={callbackBooked ? "Update" : "Set Callback"}
          disableSaveButton={callbackBooked === callbackDateTime}
          showDeleteButton={!!callbackBooked}
          onDelete={() => {
            setOpenConfirmationDialog(true);
          }}
        >
          <CallbackDialogStyles>
            <Grid item xs={12}>
              <MuiDatePicker
                showDateTimePicker
                value={callbackDateTime}
                label="Callback Date Time"
                disablePastDates
                isRequired
                handleDateChange={(value) => {
                  handleChange(value!, "callbackDateTime");
                }}
                error={highlightRequiredDetail && !callbackDateTime}
              />
            </Grid>
            {!noteOptional && (
              <Grid item xs={12} className="reason-field">
                <TextField
                  label="Reason for call"
                  value={reason}
                  onChange={(event) =>
                    handleChange(event.target.value, "reason")
                  }
                  fullWidth
                  multiline
                  rows={3}
                  required
                  error={highlightRequiredDetail && !reason}
                  helperText={
                    highlightRequiredDetail && !reason
                      ? "The reason field is required"
                      : ""
                  }
                />
              </Grid>
            )}
          </CallbackDialogStyles>
        </ResponsiveDialog>
      )}
      {openConfirmationDialog && (
        <ResponsiveDialog
          open={openConfirmationDialog}
          maxWidth="xs"
          onConfirm={() => setCallbackDetail(null)}
          onCancel={() => setOpenConfirmationDialog(false)}
          showActionButtons
          okText="Yes"
          cancelText="No"
          isConfirmationDialog
        >
          <div className="confirmation-message">
            Are you sure you want to delete the call back details?
          </div>
        </ResponsiveDialog>
      )}
    </>
  );
};
