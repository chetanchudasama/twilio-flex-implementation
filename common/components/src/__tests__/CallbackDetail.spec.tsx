/* eslint-disable react/jsx-props-no-spreading */
import { create, act } from "react-test-renderer";
import React from "react";
import { mount, shallow } from "enzyme";
import {
  CallbackDetail,
  CallbackDetailProps,
} from "../CallbackDetail/CallbackDetail";
import { CallbackDialog } from "../CallbackDialog/CallbackDialog";

const props: CallbackDetailProps = {
  setCallbackDetail: jest.fn(() => {}),
};

describe("CallbackDetail", () => {
  it("renders correctly", () => {
    const wrapper = create(<CallbackDetail {...props} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("when set callback button is clicked, should open callback detail dialog", () => {
    const wrapper = mount(<CallbackDetail {...props} />);

    expect(wrapper.find(CallbackDialog)).toHaveLength(0);

    const btn = wrapper.find(".callback-btn").at(0);
    btn.simulate("click");
    wrapper.update();

    expect(wrapper.find(CallbackDialog)).toHaveLength(1);
  });

  it("when handleCloseDialog method is called from CallbackDialog, should close callback detail dialog", () => {
    const wrapper = shallow(<CallbackDetail {...props} />);

    expect(wrapper.find(CallbackDialog)).toHaveLength(0);

    const btn = wrapper.find(".callback-btn").at(0);
    btn.simulate("click");
    wrapper.update();

    act(() => {
      wrapper.find(CallbackDialog).props().handleCloseDialog();
    });
    wrapper.update();

    expect(wrapper.find(CallbackDialog)).toHaveLength(0);
  });

  it("when update button is clicked from CallbackDialog, should emit setCallbackDetail method", () => {
    const wrapper = shallow(<CallbackDetail {...props} />);

    expect(wrapper.find(CallbackDialog)).toHaveLength(0);

    const btn = wrapper.find(".callback-btn").at(0);
    btn.simulate("click");
    wrapper.update();

    act(() => {
      wrapper.find(CallbackDialog).props().setCallbackDetail(new Date(), "");
    });
    wrapper.update();

    expect(props.setCallbackDetail).toHaveBeenCalled();
  });

  it("when value of callbackBooked prop is null, should render set callback button in UI", () => {
    const wrapper = create(<CallbackDetail {...props} />).root;

    expect(wrapper.findByProps({ className: "callback-btn" })).toBeTruthy();
  });

  it("when value of callbackBooked prop is not null, UI changed accordingly", () => {
    props.callbackBooked = new Date();
    const wrapper = create(<CallbackDetail {...props} />).root;

    expect(wrapper.findByProps({ className: "callback-booked" })).toBeTruthy();
  });
});
