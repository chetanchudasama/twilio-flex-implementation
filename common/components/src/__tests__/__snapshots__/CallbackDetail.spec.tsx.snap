// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallbackDetail renders correctly 1`] = `
<div
  className="css-12ot580"
>
  <div
    className=""
  >
    <button
      className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-outlined-9 callback-btn"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-2"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-30 MuiIcon-fontSizeSmall-37 callback-icon"
        >
          phone_enabled
        </span>
        Set Callback
      </span>
      <span
        className="MuiTouchRipple-root-39"
      />
    </button>
  </div>
</div>
`;
